keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -120,91 +120,106 @@ read/watched by different components.
 
-With those (very local and small) changes, we will introduce an internal type
-in package `cacher` implementing both `runtime.Object` and `CustomEncoder`
+With those (relatively mechanical) changes, we will introduce an internal type
+in package `cacher` implementing both `runtime.Object` and `CacheableObject`
 interfaces. The idea behind it is that it will be encapsulating the original
 object and additionally it will be able to accumulate its serialized versions.
 It will look like this:
-```
-// TODO: Better name is welcome :)
-type CachingObject struct {
-	// Object is the object (potentially in the internal version)
-	// for which serializations should be cached for future reuse.
-	Object runtime.Object
-
-	// FIXME: We may want to change that during performance experiments
-	// e.g. to use sync.Map or slice or something different, also to
-	// allow some fast-path.
-	lock sync.Mutex
-	Versioned map[runtime.GroupVersioner]*CachingVersionedObject
-}
-
-// TODO: Better name is welcome :)
-type CachingVersionedObject struct {
-	// Ojbect is the versioned object for which serializations
-	// should be cached.
-
-	// FIXME: We may want to change that during performance experiments
-	// e.g. to use sync.Map or slice or something different, also to
-	// allow some fast-path.
-	lock sync.Mutex
-	serialization map[cachingKey]*serializationResult
-}
-
-type cachineKey struct {
-	// encoder is a proxy for mediaType - given we don't have access to it
-	// (Serializer interface doesn't expose it) and we want to minimize
-	// changes to apimachinery, we take runtime.Encoder which is a singleton
-	// for a given mediaType in apiserver.
-	encoder runtime.Encoder
-	// selfLink of the serialized object identifying endpoint
-	// (e.g. group, version, subresource)
-	selfLink string
-}
-
-// TODO: Better name is welcome :)
+```go
+// serializationResult captures a result of serialization.
 type serializationResult struct {
+	// once should be used to ensure serialization is computed once.
 	once sync.Once
 
 	// raw is serialized object.
 	raw []byte
 	// err is error from serialization.
 	err error
 }
+
+// metaRuntimeInterface implements runtime.Object and
+// metav1.Object interfaces.
+type metaRuntimeInterface interface {
+	runtime.Object
+	metav1.Object
+}
+
+// cachingObject is an object that is able to cache its serializations
+// so that each of those is computed exactly once.
+//
+// cachingObject implements the metav1.Object interface (accessors for
+// all metadata fields). However, setters for all fields except from
+// SelfLink (which is set lately in the path) are ignored.
+type cachingObject struct {
+	lock sync.RWMutex
+
+	// Object for which serializations are cached.
+	object metaRuntimeInterface
+
+	// serializations is a cache containing object`s serializations.
+	// The value stored in atomic.Value is of type serializationsCache.
+	// The atomic.Value type is used to allow fast-path.
```

Comment:
What type is stored in the atomic.Value?

Reply:
The line above is saying "The value stored in atomic.Value is of type serializationsCache."

Apparently I forgot to copy the serializationsCache type from the PR:
```
// serializationsCache is a type for caching serialization results.
type serializationsCache map[runtime.Identifier]*serializationResult
```